fix: assertion log limit

sgoossens opened the pull request:
#### Purpose (TL;DR) - mandatory

Allow for a log limit setting on the assertion object, that when true, will truncate the error.message that is returned via a failing assertion.  The limit for the log message will be set to a default if none was provided

https://github.com/sinonjs/sinon/issues/2484


 #### Background (Problem in detail)  - optional
When a stubbed method has a large property on it, it is logged in the failing assertion causing runoff in the terminal and making it hard to see what assertion actually failed.

<!--
 #### Solution  - optional
-->
By allowing for a log limitation setting and a default log limit, error logs can be truncated.

#### How to verify - mandatory

1. use [this app](https://github.com/sgoossens/SinonOverflowExample) and point the sinon instance to this version
2. see the logs are reduced when the setting is on

#### Checklist for author

- [x] `npm run lint` passes
- [x] References to standard library functions are [cached](https://github.com/sinonjs/sinon/pull/1523).

fatso83 reviewed (changes requested):
Missing test coverage. ✌️

fatso83 commented on lib/sinon/assert.js:
```diff
@@ -15,8 +15,11 @@ var forEach = arrayProto.forEach;
+
+function createAssertObject(shouldLimitAssertionLogs) {
     var assert;
+    var assertionLogLimit;
```
```suggestion
```
This only makes sense to keep if you expose a way of setting it. So in the current context, it could be removed, but I assume it has been added to allow for testing this, which would 

1. Fix the only remaining issue
2. Make this line of code meaningful 😄 

P.S. We [allow newer EcmaScript syntax](https://github.com/sinonjs/sinon/blob/main/COMPATIBILITY.md#es2017), so you mist as well use `const`.

fatso83 commented:
Would you be interested in help getting it over the finish line? Have some time these days.

sgoossens commented:
> Would you be interested in help getting it over the finish line? Have some time these days.

Sure that would be great. I lost track of this but think it would be a great add. I can also put some time in to addressing your feedback as well

fatso83 commented:
OK, so I added test coverage, fixed up some logical errors, exposed the options on the sandbox creation, etc. 

This is ready for merge from my perspective, but it does not change the default, as that would be a breaking change, but we [can do that with version 18 ](https://github.com/sinonjs/sinon/milestone/10). It still allows creating sandboxes with log limits, which is probably what you want right now.

The way I stubbed out

fatso83 commented:
@sgoossens You ok with this? Anything to add/change?

sgoossens reviewed (commented):
Looks great thank you for helping get this over the line

mroderick commented on lib/sinon/assert.js:
```diff
@@ -333,4 +333,20 @@ function createAssertObject(opts) {
+
+// allow for easy mocking
+module.exports.mock = {};
+Object.defineProperty(module.exports.mock, "createAssertObject", {
```
I'm not a fan of this approach, as it adds test code to production code.

For a testing library, adding some bloat might not be that big a deal ... but for production code, especially code that ships over the internet, this adds a lot of bloat of unused code. (No, tree shaking is not a good argument to embrace this practice).

Additionally, if the code enabling tests is not added very carefully, it becomes entangled with the production code.



Since https://github.com/sinonjs/sinon/commit/11a8e3f24ccec981a280dbe63df316d1fbde4ddd, where proxyquire was removed, this seems to be one solution ... are there others?

I wish we had at least had a discussion about proxyquire and prettier conflict, so we could have evaluated solutions before they got merged into `main`.

fatso83 replied:
👍 @mroderick . I got rid of the `mock` object. It was not necessary at all, as I got this a bit confused with ESM exports for a while, and seeing it work made me realize the whole getter thing was just a totally unneccessary

fatso83 commented on test/create-sandbox-test.js:
```diff
@@ -0,0 +1,31 @@
+const createSandbox = require("../lib/sinon/create-sandbox");
+
+describe("create-sandbox", function () {
+    it("passes on options to the creation of the assert object", function () {
```
Good call. I'll fix.

fatso83 replied:
Done! Much better to keep the implementation specifics out of it.
